src/editor.rs

```diff
@@ -0,0 +1,182 @@
+    // install. If this fails, invoke the default editor instead.
+    //
+    // XXX: If we end up with commands which invoke the editor many times, it
+    // may eventually be worth adding some form of caching here.
+    match (git_sh_path(), git_editor()) {
+        (Some(git_sh), Some(git_editor)) => {
+            let mut cmd = Command::new(git_sh);
+            cmd.arg("-c")
+                .arg(format!("{} \"$@\"", git_editor))
+                .arg(git_editor);
+            return cmd;
+        }
+        (_, None) => {
+            warn!("Unable to determine user's GIT_EDITOR");
+        }
+        (None, Some(_)) => {
+            warn!("Unable to locate user's git install to invoke GIT_EDITOR");
+        }
+    }
+    warn!("Falling back to running '{}' directly", FALLBACK_EDITOR);
+    Command::new(FALLBACK_EDITOR)
+}
+
+/// Run the default editor configured through git (GIT_EDITOR) and use it to
+/// edit the given file path.
+pub fn run_editor(path: &Path) -> io::Result<ExitStatus> {
+    editor_command().arg(path).status()
+}
+
+// On windows some editors (notably notepad pre-windows 11) don't handle
+// unix line endings very well, so make sure to give them windows line
+// endings.
+#[cfg(windows)]
+const LINE_ENDING: &str = "\r\n";
+
+#[cfg(not(windows))]
+const LINE_ENDING: &str = "\n";
+
+pub struct Editor {
+    tempfile: NamedTempFile,
+}
+
+impl Editor {
+    /// Create a new editor for a temporary file.
+    pub fn new(name: &str) -> Result<Self, VetError> {
+        let tempfile = tempfile::Builder::new().prefix(name).tempfile()?;
+        Ok(Editor { tempfile })
+    }
+
+    /// Add comment lines to the editor. Any newlines in the input will be
+    /// normalized to the current platform, and a comment character will be
+    /// added.
+    pub fn add_comments(&mut self, text: &str) -> io::Result<()> {
+        let text = text.trim();
+        if text.is_empty() {
+            write!(self.tempfile, "#{}", LINE_ENDING)?;
+        }
+        for line in text.lines() {
+            if line.is_empty() {
+                write!(self.tempfile, "#{}", LINE_ENDING)?;
+            } else {
+                write!(self.tempfile, "# {}{}", line, LINE_ENDING)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Add non-comment lines to the editor. These lines must not start with a
+    /// `#` character.
+    pub fn add_text(&mut self, text: &str) -> io::Result<()> {
+        let text = text.trim();
+        if text.is_empty() {
+            write!(self.tempfile, "{}", LINE_ENDING)?;
+        }
+        for line in text.lines() {
+            // FIXME: Git has multiple comment character modes, including
+            // buffering the entire payload to pick a comment character not in
+            // the text, and respecting the user's core.commentChar config.
+            assert!(
+                !line.starts_with('#'),
+                "Non-comment lines cannot start with a '#' comment character"
+            );
```

Comment:
Yeah, I didn't love this as a failure case either. In git there's support for trying a sequence of comment characters and only falling over if all are used at the same time (https://github.com/git/git/blob/9c897eef06347cc5a3eb07c3ae409970ab1052c8/builtin/commit.c#L669-L697), which I could add but it would require some changes to buffer the file in memory and only write it out with comments once it's been fully written.

The only place where this can come up right now though is if you pass a note on the command line which starts with a `#` character on some line, as we write the note from the command line out in the file for editing.